core/commonMain/src/implementations/immutableMap/TrieNode.kt

```diff
@@ -194,28 +201,14 @@ internal class TrieNode<K, V>(
     }
 
     /** The given [newNode] must not be a part of any persistent map instance. */
-    private fun mutableUpdateNodeAtIndex(nodeIndex: Int, positionMask: Int, newNode: TrieNode<K, V>, owner: MutabilityOwnership): TrieNode<K, V> {
+    private fun mutableUpdateNodeAtIndex(nodeIndex: Int, newNode: TrieNode<K, V>, owner: MutabilityOwnership): TrieNode<K, V> {
 //        assert(buffer[nodeIndex] !== newNode)
 
-        val newNodeBuffer = newNode.buffer
-        if (newNodeBuffer.size == 2 && newNode.nodeMap == 0) {
```

Comment:
Here canonicalisation of trie can change order of iteration (see below), therefore I removed canonicalisation for mutable maps as the update may be made by an iterator currently iterating the map builder.

-------------------------               iterator.remove() after                           --------------------------
| k1 | v1 | k3 | v3 | child1 |           iterator.next() returned (k4, v4)               | k1 | v1 | k2 | v2 | k3 | v3 |
--------------------|----                                                                                --------------------------
                                  |                                     -->
             ------------------
             | k4 | v4 | k2 | v2 |
             ------------------